File: clients/algoliasearch-client-java/algoliasearch/src/main/java/com/algolia/transport/interceptors/AuthInterceptor.java
package com.algolia.transport.interceptors;

import java.io.IOException;
import okhttp3.Interceptor;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;

public final class AuthInterceptor implements Interceptor {

  private final String applicationId;
  private final String apiKey;

  public AuthInterceptor(String applicationId, String apiKey) {
    this.applicationId = applicationId;
    this.apiKey = apiKey;
  }

  @NotNull
  @Override
  public Response intercept(Chain chain) throws IOException {
    okhttp3.Request originalRequest = chain.request();
    okhttp3.Request.Builder builder = originalRequest.newBuilder();
    builder.header("x-algolia-application-id", applicationId);
    builder.header("x-algolia-api-key", apiKey);
    okhttp3.Request newRequest = builder.build();
    return chain.proceed(newRequest);
  }
}

File: clients/algoliasearch-client-scala/src/main/scala/algoliasearch/internal/interceptor/AuthInterceptor.scala
package algoliasearch.internal.interceptor

import okhttp3.{Interceptor, Response}

/** Interceptor that adds the application ID and API key to the request headers.
  *
  * @param applicationId
  *   application ID
  * @param apiKey
  *   API key
  */
private[algoliasearch] class AuthInterceptor(
    applicationId: String,
    private var apiKey: String
) extends Interceptor {

  private val HeaderApplicationId = "x-algolia-application-id"
  private val HeaderApiKey = "x-algolia-api-key"

  def setApiKey(newApiKey: String): Unit = {
    apiKey = newApiKey
  }

  override def intercept(chain: Interceptor.Chain): Response = {
    val originalRequest = chain.request()
    val builder = originalRequest.newBuilder()
    val headers = originalRequest.headers()

    if (headers.get(HeaderApplicationId) == null) {
      builder.header(HeaderApplicationId, applicationId)
    }
    if (headers.get(HeaderApiKey) == null) {
      builder.header(HeaderApiKey, apiKey)
    }

    val newRequest = builder.build()
    chain.proceed(newRequest)
  }
}
